chaincode/src/launchpad/fees.ts

```diff
@@ -0,0 +1,107 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  FeeReceiptStatus,
+  LaunchpadFeeConfig,
+  LaunchpadSale,
+  SlippageToleranceExceededError
+} from "@gala-chain/api";
+import BigNumber from "bignumber.js";
+
+import { writeChannelPaymentReceipt } from "../fees/writeChannelPaymentReceipt";
+import { writeUserPaymentReceipt } from "../fees/writeUserPaymentReceipt";
+import { transferToken } from "../transfer/transferToken";
+import { GalaChainContext } from "../types";
+import { getObjectByKey, txUnixTimeToDateIndexKeys } from "../utils";
+
+const REVERSE_BONDING_CURVE_FEE_ALPHA = new BigNumber("0.5"); // 50% of the amount of native token to be received
+const REVERSE_BONDING_CURVE_FEE_CODE = "LaunchpadReverseBondingCurveFee";
+const NATIVE_TOKEN_DECIMALS = 8;
+
+export function calculateReverseBondingCurveFee(sale: LaunchpadSale, nativeTokensToReceive: BigNumber) {
+  if (
+    !sale.reverseBondingCurveConfiguration ||
+    sale.reverseBondingCurveConfiguration.maxFeePortion.isZero()
+  ) {
+    return BigNumber(0);
+  }
+
+  const { minFeePortion, maxFeePortion } = sale.reverseBondingCurveConfiguration;
+  const feePortionDiff = maxFeePortion.minus(minFeePortion);
+  const adjustedAlpha = REVERSE_BONDING_CURVE_FEE_ALPHA.multipliedBy(feePortionDiff);
+  const feePortion = minFeePortion.plus(adjustedAlpha);
+  const circulatingSupplyProportional = sale.fetchCirculatingSupplyProportional();
+
+  const feeAmount = circulatingSupplyProportional
+    .multipliedBy(feePortion)
+    .multipliedBy(nativeTokensToReceive)
+    .decimalPlaces(NATIVE_TOKEN_DECIMALS, BigNumber.ROUND_UP);
+
+  return feeAmount;
+}
+
+export async function payReverseBondingCurveFee(
+  ctx: GalaChainContext,
+  sale: LaunchpadSale,
+  nativeTokensToReceive: BigNumber,
+  maxAcceptableFee?: BigNumber
+) {
+  const feeAmount = await calculateReverseBondingCurveFee(sale, nativeTokensToReceive);
+
+  if (feeAmount.isZero()) {
+    return; // No fee
+  }
+
+  if (maxAcceptableFee && feeAmount.isGreaterThan(maxAcceptableFee)) {
+    throw new SlippageToleranceExceededError("Fee exceeds maximum acceptable amount");
+  }
+
+  const launchpadConfigKey = ctx.stub.createCompositeKey(LaunchpadFeeConfig.INDEX_KEY, []);
+  const launchpadConfig = await getObjectByKey(ctx, LaunchpadFeeConfig, launchpadConfigKey);
+  const nativeToken = sale.fetchNativeTokenInstanceKey();
+  const { year, month, day } = txUnixTimeToDateIndexKeys(ctx.txUnixTime);
+  const txId = ctx.stub.getTxID();
+
+  await Promise.all([
```

Comment:
it's better to avoid `Promise.all` in chaincodes unless you are 100% sure you need it. In Fabric chaincodes are executed by multiple peers, and it needs to be 100% deterministic - otherwise transactions may fail. If for some reason the order of execution matters, and leads to slightly different writes, you should avoid parallel calls.

Also the optimization you get is only for reads, because writes are saved to cache with no IO required, and flushed at the end of the transaction.

Reply:
Thanks, that makes sense. I pushed a new commit to do them sequentially.